x-pack/filebeat/input/shipper/input.go

```diff
@@ -0,0 +1,279 @@
+	}
+
+	//setup gRPC
+	err := in.setupgRPC(pipeline)
+	if err != nil {
+		return fmt.Errorf("error starting shipper gRPC server: %w", err)
+	}
+	in.log.Infof("done setting up gRPC server")
+
+	// wait for shutdown
+	<-inputContext.Cancelation.Done()
+
+	in.Stop()
+
+	return nil
+}
+
+func (in *shipperInput) setupgRPC(pipeline beat.Pipeline) error {
+	in.shipperSrv = strings.TrimPrefix(in.cfg.Conn.Server, "unix://")
+	in.log.Infof("initializing grpc server at %s", in.shipperSrv)
+	// Currently no TLS until we figure out mTLS issues in agent/shipper
+	creds := insecure.NewCredentials()
+	opts := []grpc.ServerOption{
+		grpc.Creds(creds),
+		grpc.MaxRecvMsgSize(64 * units.MiB),
+	}
+
+	var err error
+	in.server = grpc.NewServer(opts...)
+	in.shipper, err = NewShipperServer(pipeline, in)
+	if err != nil {
+		return fmt.Errorf("error creating shipper server: %w", err)
+	}
+
+	pb.RegisterProducerServer(in.server, in.shipper)
+
+	in.srvMut.Lock()
+
+	// paranoid checking, make sure we have the base directory.
+	dir := filepath.Dir(in.shipperSrv)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		err = os.MkdirAll(dir, 0o755)
+		if err != nil {
+			in.srvMut.Unlock()
+			return fmt.Errorf("could not create directory for unix socket %s: %w", dir, err)
+		}
+	}
+
+	// on linux, net.Listen will fail if the file already exists
+	if _, err = os.Stat(in.shipperSrv); err == nil {
+		in.log.Debugf("listen address %s already exists, removing", in.shipperSrv)
+		err = os.Remove(in.shipperSrv)
+		if err != nil {
+			in.srvMut.Unlock()
+			return fmt.Errorf("error removing unix socket %s: %w", in.shipperSrv, err)
+		}
+	}
+
+	lis, err := newListener(in.log, in.shipperSrv)
+	if err != nil {
+		in.srvMut.Unlock()
+		return fmt.Errorf("failed to listen on %s: %w", in.shipperSrv, err)
+	}
+
+	go func() {
+		in.log.Infof("gRPC listening on %s", in.shipperSrv)
+		err = in.server.Serve(lis)
+		if err != nil {
+			in.log.Errorf("gRPC server shut down with error: %s", err)
+		}
+	}()
+
+	// make sure connection is up before mutex is released
+	defer in.srvMut.Unlock()
+	con, err := tools.DialTestAddr(in.shipperSrv)
+	if err != nil {
+		// this will stop the other go routine in the wait group
+		in.server.Stop()
+		return fmt.Errorf("failed to test connection with the gRPC server on %s: %w", in.shipperSrv, err)
+	}
+	_ = con.Close()
+
+	return nil
+}
+
+func (in *shipperInput) sendEvent(event *messages.Event) (queue.EntryID, error) {
+	//look for matching processor config
+	stream, ok := in.streams[event.Source.StreamId]
+	// should this be an error? can we continue on if there's no data stream associated with an event?
+	if !ok {
+		return queue.EntryID(0), fmt.Errorf("could not find data stream associated with ID '%s'", event.Source.StreamId)
+	}
+
+	// inject index from stream config into metadata
+	// not sure how stock beats do this, and also not sure if it's actually needed?
+	metadata := helpers.AsMap(event.Metadata)
+	metadata["index"] = stream.index
```

Comment:
we shouldn't need this.  I think the data_stream is already in the `Fields` but if not we should just add the data_stream, not index.

Reply:
Yah, that's what I thought too, but I wasn't 100% sure. Although not sure how we should attempt to add `data_stream` values if they don't already exist?

Reply:
The Event stuct has Datastream.  https://github.com/elastic/elastic-agent-shipper-client/blob/da2edfba0431acad015a03e8f184b970ba3457b6/pkg/proto/messages/publish.pb.go#LL96C1-L111C2

Clients have to fill this in.

Reply:
Yah, I assume that would be client-side. Your first comment made me think that the shipper should somehow fill in the missing `Datastream` object if it can't find one.